Use self-hosted agent

leofang opened the pull request:
While waiting for the CI let me lay out my thoughts and plan... 

The upstream recipe is maintained by John Kirkham and me. But the most challenging part is that we do not have access to GPU resources for testing the recipe, so we've hit hard by some bugs along the way. Since now we have our own clone of the CF build chain, and
1. our ptychography software needs CuPy,
2. we are exploring the possibility of using local machines to build some packages,

it makes perfect sense to use this feedstock as a testbed/playground.

Moreover, as a CuPy developer I would love to build development versions for local benchmarking purposes. This is something CF's policy does not allow but I would love to see happening. If everything works out fine and resource permits, I wanna build
- ~~v7.3.0 (this PR)~~
- v7.4.0 **(UPDATE: this PR)**
- any future official v7 releases
- v8.0.0a1 
- v8.0.0b1
- v8.0.0b2 (just released)
- any development and official releases of v8

However there're some technical issues I'll have to address first.

leofang committed:
use nsls2forge Azure org

leofang committed:
copy CF data

leofang committed:
rerendered

leofang committed:
reduce build matrix for testing; rerendered

leofang committed:
use local runner for CI; DO NOT rerender!

leofang commented:
790b71c is all we need to switch to a local agent!

leofang commented:
It builds but the test says GPU not found...Looking into this locally...

leofang committed:
print debug info as I can't reproduce the failure locally

copied from conda-forge/cupy-feedstock#8. don't think this would help
much, just give it a shot...

leofang commented:
CUDA driver is not loaded in the docker image. It seems we can't use CF's toolchain and instead need to define our own Azure workflow as in NSLS-II/ptycho_gui#90...?

leofang committed:
try nvidia-docker

leofang commented on .scripts/run_docker_build.sh:
```diff
@@ -59,7 +59,7 @@ if [ -z "${CI}" ]; then
 
 export UPLOAD_PACKAGES="${UPLOAD_PACKAGES:-True}"
-docker run ${DOCKER_RUN_ARGS} \
+nvidia-docker run ${DOCKER_RUN_ARGS} \
```
This is a machine-specific change that makes the GPUs recognized by docker. `nvidia-docker` is discontinued as NVIDIA drivers are now natively supported by docker, but on our machine somehow it has some problems which I haven't diagnosed, and I noticed the old `nvidia-docker` still works.

leofang replied:
As docker is heavily used on this machine, I prefer not to fix/update docker for now.

leofang reviewed (commented):
Nice! Test suites are running on GPU! 🎉

leofang committed:
set docker executable properly; fix TTY issue; re-enabled CUB; rerendered

The TTY issue gives this message:
bash: cannot set terminal process group (-1): Inappropriate ioctl for device
bash: no job control in this shell

leofang committed:
fix TTY issue (retry)

leofang committed:
Azure: clean workspace before building

leofang committed:
run all tests

leofang committed:
try adding compilers

leofang committed:
skip install tests

leofang commented on recipe/meta.yaml:
```diff
@@ -64,6 +68,10 @@ test:
+    - {{ compiler("c") }}
+    - {{ compiler("cxx") }}
+    - {{ compiler("cuda") }}  # tests need nvcc
+
```
```suggestion
    - scipy

```

leofang commented on recipe/run_test.py:
```diff
@@ -0,0 +1,22 @@
+import py
+py.test.cmdline.main(["tests/cupy_tests"])
+py.test.cmdline.main(["tests/cupyx_tests"])
+py.test.cmdline.main(["tests/example_tests"])
```
```suggestion

```

leofang committed:
skip example tests; test against scipy

leofang committed:
test more CUDA versions; rerendered

leofang committed:
rerendered with conda-forge/conda-smithy#1296

Once that PR is merged, no manual intervention is needed!

leofang commented:
Manually editing `.azure-pipelines/azure-pipelines-linux.yml` can be avoided with conda-forge/conda-smithy#1296!

leofang committed:
print runtime info

jakirkham commented:
Thanks for working on this Leo! If we are able to get this to run using a self-hosted agent, what do you think about adding this to the conda-forge feedstock?

leofang committed:
try more explicit pinning for scipy; rerendered with https://github.com/conda-forge/conda-smithy/pull/1296/commits/827819f6e0bd337751bc5859a84aa6801f5c87e1

leofang commented:
Someone has to pay for the GPU resources 😉 In our case this is 1. for my own curiosity and 2. exploring and prepping for future internal needs, possibly involving closed-source and/or patent packages. (Btw I am not sure whether this feedstock should survive, will need to consult @mrakitin after his return. Also, we will migrate to other machines after lockdown is lifted.)

I am not the person who can make the final call (nor would I like to), but my understanding is that some regulations would prohibit our (federal) machines from being exposed to somewhere we don't have sole control/ownership (e.g. CF), so at least it's not possible with this machine. I will be more than happy to set the agents up for CF if access to some resources is granted to me. 

Is it possible with gpuCI? I've been wondering about this for a while...

leofang committed:
remove scipy pinning (again)

leofang committed:
rerendered with https://github.com/conda-forge/conda-smithy/pull/1296/commits/ed54f8c49b719066a1936ba9f015e6fbac0583de

leofang committed:
update to v7.4.0 and sync with upstream

leofang committed:
build cutensor support and bundle cutensor

leofang committed:
TEST: import cutensor

leofang commented:
https://github.com/conda-forge/conda-smithy/pull/1296 is merged 🎉

leofang committed:
undo cutensor support; enlarge build matrix; rerendered with conda-smithy master

The issue with cuTENSOR is brought up to upstream (conda-forge/cupy-feedstock/#57)

leofang committed:
support both PY36 and PY37

leofang commented:
The cuTENSOR support is not built due to https://github.com/conda-forge/cupy-feedstock/issues/57.

mrakitin commented:
> [conda-forge/conda-smithy#1296](https://github.com/conda-forge/conda-smithy/pull/1296) is merged 🎉

Yay! Congrats :)

leofang commented:
Tests are passed. Let me self-merge and decide what to do next.

mrakitin commented:
Sounds good to me.
